src/PrestaShopBundle/Controller/Admin/Improve/Payment/PaymentPreferencesController.php

```diff
@@ -89,21 +100,92 @@ public function indexAction(Request $request)
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentCountryRestrictionsFormAction(Request $request): Response
+    {
+        $countryRestrictionsFormHandler = $this->getPaymentCountryRestrictionsFormHandler();
 
-            $errors = $paymentPreferencesFormHandler->save($paymentPreferences);
-            if (empty($errors)) {
-                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        return $this->processForm($countryRestrictionsFormHandler, $request);
+    }
 
-                return $this->redirectToRoute('admin_payment_preferences');
+    /**
+     * Process payment modules preferences currency restrictions form.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentCurrencyRestrictionsFormAction(Request $request): Response
+    {
+        $currencyRestrictionsFormHandler = $this->getPaymentCurrencyRestrictionsFormHandler();
+
+        return $this->processForm($currencyRestrictionsFormHandler, $request);
+    }
+
+    /**
+     * Process payment modules preferences group restrictions form.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentGroupRestrictionsFormAction(Request $request): Response
+    {
+        $groupRestrictionsFormHandler = $this->getPaymentGroupRestrictionsFormHandler();
+
+        return $this->processForm($groupRestrictionsFormHandler, $request);
+    }
+
+    /**
+     * Processes the form in a generic way.
+     *
+     * @param FormHandlerInterface $formHandler
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    private function processForm(FormHandlerInterface $formHandler, Request $request): RedirectResponse
+    {
+        $form = $formHandler->getForm();
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted()) {
+            $data = $form->getData();
+
+            $errors = $formHandler->save($data);
+            if (empty($errors)) {
+                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
             }
 
             $this->flashErrors($errors);
```

Comment:
```suggestion
            $errors = $formHandler->save($data);
            if (empty($errors)) {
                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
            } else {
                $this->flashErrors($errors);
            }
```

I'm not sure it's necessary to run falshErrors method without having an errors inside :)

Reply:
Good point!